api/configuration/v1alpha1/kongkeyset_types.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2024 Kong, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	konnectv1alpha1 "github.com/kong/kubernetes-configuration/api/konnect/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// KongKeySet is the schema for KongKeySet API which defines a KongKeySet entity.
+//
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Programmed",description="The Resource is Programmed on Konnect",type=string,JSONPath=`.status.conditions[?(@.type=='Programmed')].status`
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.spec.controlPlaneRef) || has(self.spec.controlPlaneRef)", message="controlPlaneRef is required once set"
+// +kubebuilder:validation:XValidation:rule="(!self.status.conditions.exists(c, c.type == 'Programmed' && c.status == 'True')) ? true : oldSelf.spec.controlPlaneRef == self.spec.controlPlaneRef", message="spec.controlPlaneRef is immutable when an entity is already Programmed"
+// +kubebuilder:validation:XValidation:rule="!has(self.spec.controlPlaneRef) ? true : !has(self.spec.controlPlaneRef.konnectNamespacedRef) ? true : !has(self.spec.controlPlaneRef.konnectNamespacedRef.__namespace__)", message="spec.controlPlaneRef cannot specify namespace for namespaced resource - it's not supported yet"
+type KongKeySet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec KongKeySetSpec `json:"spec"`
+
+	// +kubebuilder:default={conditions: {{type: "Programmed", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	Status KongKeySetStatus `json:"status,omitempty"`
+}
+
+// KongKeySetSpec defines the spec for a KongKeySet.
+type KongKeySetSpec struct {
+	// ControlPlaneRef is a reference to a Konnect ControlPlane with which KongKeySet is associated.
+	// +optional
+	ControlPlaneRef *ControlPlaneRef `json:"controlPlaneRef,omitempty"`
```

Comment:
Since this is a Konnect focused type and #31 doesn't apply here and because of the fact that https://kong-platform-api.netlify.app/konnect/control-planes-config/v2/openapi.yaml/#tag/KeySets/operation/create-key-set does mark the CP ref as required I believe we should require it here as well.

Reply:
Hmm, it makes sense. It's a net new type that no one has used so we can make the CP ref required. That will imply that if we ever decide to support it in KIC, it will only be supported with the explicit KIC CP ref.

Reply:
Hm there's still KIC's use case as you've pointed out. I'm wondering what approach we should take here: what to do with CP refs in CRDs that are not going to be used in KIC. So far I we've followed the path of making this optional (just like you have done in this PR) which allows KIC's use case with CP ref type being of KIC type by default.

Having said that I believe we can move forward with this. WDYT?

Reply:
For sure making it optional makes our approach consistent across all CRDs. If you're ok with that, I can revert 
1c4d039 changes and leave it as was.

Reply:
d90226d

Reply:
I believe it is for the better to keep consistent now.